module/zfs/ddt.c

```diff
@@ -816,14 +816,14 @@ ddt_entry_compare(const void *x1, const void *x2)
 {
 	const ddt_entry_t *dde1 = x1;
 	const ddt_entry_t *dde2 = x2;
-	const uint64_t *u1 = (const uint64_t *)&dde1->dde_key;
-	const uint64_t *u2 = (const uint64_t *)&dde2->dde_key;
+	const ddt_key_cmp_t *u1 = (ddt_key_cmp_t *)&dde1->dde_key;
+	const ddt_key_cmp_t *u2 = (ddt_key_cmp_t *)&dde2->dde_key;
 	int i;
 
 	for (i = 0; i < DDT_KEY_WORDS; i++) {
-		if (u1[i] < u2[i])
+		if (u1->dkc_key[i] < u2->dkc_key[i])
 			return (-1);
-		if (u1[i] > u2[i])
+		if (u1->dkc_key[i] > u2->dkc_key[i])
 			return (1);
 	}
```

Comment:
I'm sorry, but I don't quite follow you, why pick the compare of dkc_key[0] out of the for loop? 

BTW, since the common case is unequal as you say, and we can save 3 comparisons only when the 256-bit checksum compeletely the same, is it necessary enough to increase the complexity here?